src/lavinmq/queue/stream_queue.cr

```diff
@@ -0,0 +1,187 @@
+require "./durable_queue"
+
+module LavinMQ
+  class StreamQueue < Queue
+    @durable = true
+    @exclusive_consumer = false
+    @no_ack = true
+
+    private def init_msg_store(data_dir)
+      @msg_store = StreamQueueMessageStore.new(data_dir)
+    end
+
+    class StreamQueueMessageStore < MessageStore
+      def initialize(@data_dir : String)
+        super
+        # message id? segment position?
+      end
+
+      def shift? : Envelope? # ameba:disable Metrics/CyclomaticComplexity
+        raise ClosedError.new if @closed
+        if sp = @requeued.shift?
+          segment = @segments[sp.segment]
+          begin
+            msg = BytesMessage.from_bytes(segment.to_slice + sp.position)
+            @bytesize -= sp.bytesize
+            @size -= 1
+            notify_empty(true) if @size.zero?
+            return Envelope.new(sp, msg, redelivered: true)
+          rescue ex
+            raise Error.new(segment, cause: ex)
+          end
+        end
+
+        loop do
+          rfile = @rfile
+          seg = @rfile_id
+          pos = rfile.pos.to_u32
+          if pos == rfile.size # EOF?
+            select_next_read_segment && next
+            return if @size.zero?
+            raise IO::EOFError.new("EOF but @size=#{@size}")
+          end
+          if deleted?(seg, pos)
+            BytesMessage.skip(rfile)
+            next
+          end
+          msg = BytesMessage.from_bytes(rfile.to_slice + pos)
+          sp = SegmentPosition.make(seg, pos, msg)
+          rfile.seek(sp.bytesize, IO::Seek::Current)
+          @bytesize -= sp.bytesize
+          @size -= 1
+          notify_empty(true) if @size.zero?
+          return Envelope.new(sp, msg, redelivered: false)
+        rescue ex
+          raise Error.new(@rfile, cause: ex)
+        end
+      end
+
+      # should delete without ack
+      def delete(sp) : Nil
+      end
+    end
+
+    def add_offset_header(msg, pos)
+      headers = msg.properties.headers || ::AMQP::Client::Arguments.new
+      headers["x-stream-offset"] = pos.as(AMQ::Protocol::Field)
+      msg.properties.headers = headers
+      msg
+    end
+
+    # save message id / segment position
+    def publish(msg : Message) : Bool
```

Comment:
i would suggest adding the SegmentPosition as the offset, but do it in the MessageStore, not here

Reply:
I don't think we can/should use SegmentPositions as offset, because their position depend on size of message and so on, right? 

For streams in Rabbit, the offset is just an incrementing id, so you always know that the next message after n will have offset n+1. And when you start consuming with an offset=n, you get the message with offset=n, so you need to be able to figure out the offset of the next message.